cmd/cluster/scale/defaulting/scaling.go

```diff
@@ -0,0 +1,131 @@
+package defaulting
+
+import (
+	"context"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/ruleengine"
+
+	"github.com/giantswarm/gsctl/cmd/cluster/scale/request"
+)
+
+type ScalingConfig struct {
+	AutoScalingEnabled       *bool
+	CurrentScalingMax        *int64
+	CurrentScalingMin        *int64
+	DesiredNumWorkers        *int64
+	DesiredNumWorkersChanged *bool
+	DesiredScalingMax        *int64
+	DesiredScalingMaxChanged *bool
+	DesiredScalingMin        *int64
+	DesiredScalingMinChanged *bool
+}
+
+type Scaling struct {
+	autoScalingEnabled       bool
+	currentScalingMax        int64
+	currentScalingMin        int64
+	desiredNumWorkers        int64
+	desiredNumWorkersChanged bool
+	desiredScalingMax        int64
+	desiredScalingMaxChanged bool
+	desiredScalingMin        int64
+	desiredScalingMinChanged bool
+}
+
+func NewScaling(config ScalingConfig) (*Scaling, error) {
+	if config.AutoScalingEnabled == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.AutoScalingEnabled must not be empty", config)
+	}
+
+	s := &Scaling{
+		autoScalingEnabled:       *config.AutoScalingEnabled,
+		currentScalingMax:        *config.CurrentScalingMax,
+		currentScalingMin:        *config.CurrentScalingMin,
+		desiredNumWorkers:        *config.DesiredNumWorkers,
+		desiredNumWorkersChanged: *config.DesiredNumWorkersChanged,
+		desiredScalingMax:        *config.DesiredScalingMax,
+		desiredScalingMaxChanged: *config.DesiredScalingMaxChanged,
+		desiredScalingMin:        *config.DesiredScalingMin,
+		desiredScalingMinChanged: *config.DesiredScalingMinChanged,
+	}
+
+	return s, nil
+}
+
+func (s *Scaling) Default(ctx context.Context, scaling request.Scaling) request.Scaling {
+	rules := []ruleengine.DefaultingRule{
+		{
+			Conditions: []func() bool{
+				func() bool { return s.autoScalingEnabled },
+				func() bool { return s.desiredScalingMinChanged },
+			},
+			Defaulting: func() {
+				scaling.Min = s.desiredScalingMin
+			},
+		},
+		{
+			Conditions: []func() bool{
+				func() bool { return s.autoScalingEnabled },
+				func() bool { return s.desiredScalingMaxChanged },
+			},
+			Defaulting: func() {
+				scaling.Max = s.desiredScalingMax
+			},
+		},
+		{
+			Conditions: []func() bool{
+				func() bool { return s.autoScalingEnabled },
+				func() bool { return !s.desiredScalingMinChanged },
```

Comment:
Just a note while reviewing this: Spotting difference between this block and first block was an exercise and made me initially question if there's something wrong when `scaling.Min` changes twice in a row - until I spotted `!`.

What do you think if there was a function for this kind of cases in order to improve readability? The above function could be then:
```go
func() bool { return not(s.desiredScalingMingChanged) },
```

or if we go the rabbit hole even deeper:
```go
func() bool { return when(not(s.desiredScalingMingChanged)) },
```

The helper functions could be:
```go

func not(v bool) bool {
    return !v
}

func when(v bool) bool {
    return v
}
```

I'd figure the compiler optimizes these away and they would effectively just help with readability.  I'm unsure about the latter `when()` though, but just trying to think ways to make reviewing easier :slightly_smiling_face:

Reply:
We do not have optimization issues, especially not in the command line clients. Not on this level. I like the idea with `not`. This is a good idea. Not sure about `when`. I figured it is pretty hard to read with longer variable names in camel case. In the future I would like to have more structured structures so to speak. `desiredScalingMinChanged` would then be `desired.Scaling.Min.Changed`.

Reply:
I added `not`.